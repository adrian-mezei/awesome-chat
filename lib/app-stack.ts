import * as Lambda from '@aws-cdk/aws-lambda';
import * as path from 'path';
import * as cdk from '@aws-cdk/core';
import * as ApiGW from '@aws-cdk/aws-apigatewayv2';
import * as Iam from '@aws-cdk/aws-iam';
import * as S3 from '@aws-cdk/aws-s3';
import * as S3Deployment from '@aws-cdk/aws-s3-deployment';
import * as Logs from '@aws-cdk/aws-logs';
import * as fs from 'fs-extra';
import * as DDB from '@aws-cdk/aws-dynamodb';
import { RemovalPolicy } from '@aws-cdk/core';

export class AwesomeChatStack extends cdk.Stack {
    constructor(scope: cdk.Construct, id: string, props?: cdk.StackProps) {
        super(scope, id, props);

        const lambdaFunction = this.createLambdaFunction();
        const { webSocketApi, webSocketApiStage } = this.createApiGateway(lambdaFunction);

        const s3Bucket = this.createS3Bucket(webSocketApi, webSocketApiStage);

        const dynamoDBTable = this.createDynamoDBTable();

        this.updateLambdaFunction(lambdaFunction, webSocketApi, webSocketApiStage, dynamoDBTable);
    }

    createLambdaFunction(): Lambda.Function {
        const lambdaFunction = new Lambda.Function(this, 'WebSocketHandler', {
            runtime: Lambda.Runtime.NODEJS_12_X,
            handler: 'index.handler',
            code: Lambda.Code.fromAsset(path.join(__dirname, './../app/backend/bundle')),
            logRetention: Logs.RetentionDays.THREE_DAYS,
        });

        lambdaFunction.grantInvoke(new Iam.ServicePrincipal('apigateway.amazonaws.com'));

        return lambdaFunction;
    }

    updateLambdaFunction(
        lambdaFunction: Lambda.Function,
        webSocketApi: ApiGW.WebSocketApi,
        webSocketApiStage: ApiGW.WebSocketStage,
        dynamoDBTable: DDB.Table,
    ): void {
        lambdaFunction.addEnvironment(
            'APIGW_ENDPOINT',
            `https://${webSocketApi.apiId}.execute-api.${this.region}.amazonaws.com/${webSocketApiStage.stageName}`,
        );

        lambdaFunction.addToRolePolicy(
            new Iam.PolicyStatement({
                actions: ['execute-api:ManageConnections'],
                resources: [
                    `arn:aws:execute-api:${this.region}:${this.account}:${webSocketApi.apiId}/${webSocketApiStage.stageName}/POST/*`,
                ],
            }),
        );

        lambdaFunction.addEnvironment('TABLE_NAME', dynamoDBTable.tableName);

        lambdaFunction.addToRolePolicy(
            new Iam.PolicyStatement({
                actions: ['dynamodb:DeleteItem', 'dynamodb:Scan', 'dynamodb:PutItem'],
                resources: [`arn:aws:dynamodb:${this.region}:${this.account}:table/${dynamoDBTable.tableName}`],
            }),
        );

        dynamoDBTable.grantReadWriteData(lambdaFunction);
    }

    createApiGateway(
        lambdaFunction: Lambda.Function,
    ): { webSocketApi: ApiGW.WebSocketApi; webSocketApiStage: ApiGW.WebSocketStage } {
        const bind: () => ApiGW.WebSocketRouteIntegrationConfig = () => ({
            type: ApiGW.WebSocketIntegrationType.AWS_PROXY,
            uri: `arn:aws:apigateway:${this.region}:lambda:path/2015-03-31/functions/${lambdaFunction.functionArn}/invocations`,
        });
        const webSocketApi = new ApiGW.WebSocketApi(this, 'web-socket-api', {
            connectRouteOptions: { integration: { bind } },
            disconnectRouteOptions: { integration: { bind } },
            defaultRouteOptions: { integration: { bind } },
        });

        const webSocketApiStage = new ApiGW.WebSocketStage(this, 'web-socket-api-stage', {
            webSocketApi,
            stageName: 'prod',
            autoDeploy: true,
        });

        return { webSocketApi, webSocketApiStage };
    }

    createS3Bucket(webSocketApi: ApiGW.WebSocketApi, webSocketApiStage: ApiGW.WebSocketStage): S3.Bucket {
        const bucket = new S3.Bucket(this, 's3-static-website-bucket', {
            publicReadAccess: true,
            removalPolicy: cdk.RemovalPolicy.DESTROY,
            websiteIndexDocument: 'index.html',
        });

        const deployment = new S3Deployment.BucketDeployment(this, 's3-static-website-deployment', {
            sources: [S3Deployment.Source.asset(path.join(__dirname, './../app/frontend'))],
            destinationBucket: bucket,
        });

        return bucket;
    }

    createDynamoDBTable(): DDB.Table {
        return new DDB.Table(this, 'dynamodb-table', {
            partitionKey: { name: 'connectionId', type: DDB.AttributeType.STRING },
            billingMode: DDB.BillingMode.PAY_PER_REQUEST,
            removalPolicy: RemovalPolicy.DESTROY,
            tableName: 'chat-app-table',
        });
    }
}
