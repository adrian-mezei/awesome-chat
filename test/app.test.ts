import { expect as expectCDK, matchTemplate, MatchStyle, SynthUtils } from '@aws-cdk/assert';
import * as cdk from '@aws-cdk/core';
import { Stack } from '@aws-cdk/core';
import * as App from '../lib/app-stack';

test('Test Stack', () => {
    const app = new cdk.App();

    const stack = new App.AwesomeChatStack(app, 'MyTestStack');

    expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
});
