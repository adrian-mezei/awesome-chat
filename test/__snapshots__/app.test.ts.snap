// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Stack 1`] = `
Object {
  "Parameters": Object {
    "AssetParameters450574362050a6ab3f053713b81bf81caf62329157f069c8c6eb58698a03b1c2ArtifactHashC845AAEF": Object {
      "Description": "Artifact hash for asset \\"450574362050a6ab3f053713b81bf81caf62329157f069c8c6eb58698a03b1c2\\"",
      "Type": "String",
    },
    "AssetParameters450574362050a6ab3f053713b81bf81caf62329157f069c8c6eb58698a03b1c2S3BucketED4A2582": Object {
      "Description": "S3 bucket for asset \\"450574362050a6ab3f053713b81bf81caf62329157f069c8c6eb58698a03b1c2\\"",
      "Type": "String",
    },
    "AssetParameters450574362050a6ab3f053713b81bf81caf62329157f069c8c6eb58698a03b1c2S3VersionKey1E02BCAA": Object {
      "Description": "S3 key for asset version \\"450574362050a6ab3f053713b81bf81caf62329157f069c8c6eb58698a03b1c2\\"",
      "Type": "String",
    },
    "AssetParameters6153562b1c727a86db389e75e42e97be201573b884dadafaed4f58ff7b7ecd06ArtifactHash6EE160BC": Object {
      "Description": "Artifact hash for asset \\"6153562b1c727a86db389e75e42e97be201573b884dadafaed4f58ff7b7ecd06\\"",
      "Type": "String",
    },
    "AssetParameters6153562b1c727a86db389e75e42e97be201573b884dadafaed4f58ff7b7ecd06S3BucketE98F073A": Object {
      "Description": "S3 bucket for asset \\"6153562b1c727a86db389e75e42e97be201573b884dadafaed4f58ff7b7ecd06\\"",
      "Type": "String",
    },
    "AssetParameters6153562b1c727a86db389e75e42e97be201573b884dadafaed4f58ff7b7ecd06S3VersionKey54E58213": Object {
      "Description": "S3 key for asset version \\"6153562b1c727a86db389e75e42e97be201573b884dadafaed4f58ff7b7ecd06\\"",
      "Type": "String",
    },
    "AssetParameters67b7823b74bc135986aa72f889d6a8da058d0c4a20cbc2dfc6f78995fdd2fc24ArtifactHashBA91B77F": Object {
      "Description": "Artifact hash for asset \\"67b7823b74bc135986aa72f889d6a8da058d0c4a20cbc2dfc6f78995fdd2fc24\\"",
      "Type": "String",
    },
    "AssetParameters67b7823b74bc135986aa72f889d6a8da058d0c4a20cbc2dfc6f78995fdd2fc24S3Bucket4D46ABB5": Object {
      "Description": "S3 bucket for asset \\"67b7823b74bc135986aa72f889d6a8da058d0c4a20cbc2dfc6f78995fdd2fc24\\"",
      "Type": "String",
    },
    "AssetParameters67b7823b74bc135986aa72f889d6a8da058d0c4a20cbc2dfc6f78995fdd2fc24S3VersionKeyB0F28861": Object {
      "Description": "S3 key for asset version \\"67b7823b74bc135986aa72f889d6a8da058d0c4a20cbc2dfc6f78995fdd2fc24\\"",
      "Type": "String",
    },
    "AssetParametersc24b999656e4fe6c609c31bae56a1cf4717a405619c3aa6ba1bc686b8c2c86cfArtifactHash85F58E48": Object {
      "Description": "Artifact hash for asset \\"c24b999656e4fe6c609c31bae56a1cf4717a405619c3aa6ba1bc686b8c2c86cf\\"",
      "Type": "String",
    },
    "AssetParametersc24b999656e4fe6c609c31bae56a1cf4717a405619c3aa6ba1bc686b8c2c86cfS3Bucket55EFA30C": Object {
      "Description": "S3 bucket for asset \\"c24b999656e4fe6c609c31bae56a1cf4717a405619c3aa6ba1bc686b8c2c86cf\\"",
      "Type": "String",
    },
    "AssetParametersc24b999656e4fe6c609c31bae56a1cf4717a405619c3aa6ba1bc686b8c2c86cfS3VersionKey60329B70": Object {
      "Description": "S3 key for asset version \\"c24b999656e4fe6c609c31bae56a1cf4717a405619c3aa6ba1bc686b8c2c86cf\\"",
      "Type": "String",
    },
    "AssetParameterse9882ab123687399f934da0d45effe675ecc8ce13b40cb946f3e1d6141fe8d68ArtifactHashD9A515C3": Object {
      "Description": "Artifact hash for asset \\"e9882ab123687399f934da0d45effe675ecc8ce13b40cb946f3e1d6141fe8d68\\"",
      "Type": "String",
    },
    "AssetParameterse9882ab123687399f934da0d45effe675ecc8ce13b40cb946f3e1d6141fe8d68S3BucketAEADE8C7": Object {
      "Description": "S3 bucket for asset \\"e9882ab123687399f934da0d45effe675ecc8ce13b40cb946f3e1d6141fe8d68\\"",
      "Type": "String",
    },
    "AssetParameterse9882ab123687399f934da0d45effe675ecc8ce13b40cb946f3e1d6141fe8d68S3VersionKeyE415415F": Object {
      "Description": "S3 key for asset version \\"e9882ab123687399f934da0d45effe675ecc8ce13b40cb946f3e1d6141fe8d68\\"",
      "Type": "String",
    },
  },
  "Resources": Object {
    "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756C81C01536": Object {
      "DependsOn": Array [
        "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRoleDefaultPolicy88902FDF",
        "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRole89A01265",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParametersc24b999656e4fe6c609c31bae56a1cf4717a405619c3aa6ba1bc686b8c2c86cfS3Bucket55EFA30C",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParametersc24b999656e4fe6c609c31bae56a1cf4717a405619c3aa6ba1bc686b8c2c86cfS3VersionKey60329B70",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParametersc24b999656e4fe6c609c31bae56a1cf4717a405619c3aa6ba1bc686b8c2c86cfS3VersionKey60329B70",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Handler": "index.handler",
        "Layers": Array [
          Object {
            "Ref": "s3staticwebsitedeploymentAwsCliLayerAA8A2B35",
          },
        ],
        "Role": Object {
          "Fn::GetAtt": Array [
            "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRole89A01265",
            "Arn",
          ],
        },
        "Runtime": "python3.6",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRole89A01265": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRoleDefaultPolicy88902FDF": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      Object {
                        "Ref": "AssetParameters450574362050a6ab3f053713b81bf81caf62329157f069c8c6eb58698a03b1c2S3BucketED4A2582",
                      },
                    ],
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      Object {
                        "Ref": "AssetParameters450574362050a6ab3f053713b81bf81caf62329157f069c8c6eb58698a03b1c2S3BucketED4A2582",
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
            Object {
              "Action": Array [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
                "s3:DeleteObject*",
                "s3:PutObject*",
                "s3:Abort*",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "s3staticwebsitebucket87408974",
                    "Arn",
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::GetAtt": Array [
                          "s3staticwebsitebucket87408974",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRoleDefaultPolicy88902FDF",
        "Roles": Array [
          Object {
            "Ref": "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRole89A01265",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aFD4BFC8A": Object {
      "DependsOn": Array [
        "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB",
        "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParameters67b7823b74bc135986aa72f889d6a8da058d0c4a20cbc2dfc6f78995fdd2fc24S3Bucket4D46ABB5",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters67b7823b74bc135986aa72f889d6a8da058d0c4a20cbc2dfc6f78995fdd2fc24S3VersionKeyB0F28861",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters67b7823b74bc135986aa72f889d6a8da058d0c4a20cbc2dfc6f78995fdd2fc24S3VersionKeyB0F28861",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
            "Arn",
          ],
        },
        "Runtime": "nodejs12.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "logs:PutRetentionPolicy",
                "logs:DeleteRetentionPolicy",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB",
        "Roles": Array [
          Object {
            "Ref": "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "WebSocketHandler47C0AA1A": Object {
      "DependsOn": Array [
        "WebSocketHandlerServiceRoleDefaultPolicyF180304B",
        "WebSocketHandlerServiceRole79B07D0E",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParameters6153562b1c727a86db389e75e42e97be201573b884dadafaed4f58ff7b7ecd06S3BucketE98F073A",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters6153562b1c727a86db389e75e42e97be201573b884dadafaed4f58ff7b7ecd06S3VersionKey54E58213",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters6153562b1c727a86db389e75e42e97be201573b884dadafaed4f58ff7b7ecd06S3VersionKey54E58213",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Environment": Object {
          "Variables": Object {
            "APIGW_ENDPOINT": Object {
              "Fn::Join": Array [
                "",
                Array [
                  "https://",
                  Object {
                    "Ref": "websocketapi45E8C8EB",
                  },
                  ".execute-api.",
                  Object {
                    "Ref": "AWS::Region",
                  },
                  ".amazonaws.com/prod",
                ],
              ],
            },
            "TABLE_NAME": Object {
              "Ref": "dynamodbtable650E77A7",
            },
          },
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "WebSocketHandlerServiceRole79B07D0E",
            "Arn",
          ],
        },
        "Runtime": "nodejs12.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "WebSocketHandlerInvokeServicePrincipalapigatewayamazonawscomE462079C": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "WebSocketHandler47C0AA1A",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
      },
      "Type": "AWS::Lambda::Permission",
    },
    "WebSocketHandlerLogRetention7648D10D": Object {
      "Properties": Object {
        "LogGroupName": Object {
          "Fn::Join": Array [
            "",
            Array [
              "/aws/lambda/",
              Object {
                "Ref": "WebSocketHandler47C0AA1A",
              },
            ],
          ],
        },
        "RetentionInDays": 3,
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aFD4BFC8A",
            "Arn",
          ],
        },
      },
      "Type": "Custom::LogRetention",
    },
    "WebSocketHandlerServiceRole79B07D0E": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "WebSocketHandlerServiceRoleDefaultPolicyF180304B": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "execute-api:ManageConnections",
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:aws:execute-api:",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":",
                    Object {
                      "Ref": "websocketapi45E8C8EB",
                    },
                    "/prod/POST/*",
                  ],
                ],
              },
            },
            Object {
              "Action": Array [
                "dynamodb:DeleteItem",
                "dynamodb:Scan",
                "dynamodb:PutItem",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:aws:dynamodb:",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":table/",
                    Object {
                      "Ref": "dynamodbtable650E77A7",
                    },
                  ],
                ],
              },
            },
            Object {
              "Action": Array [
                "dynamodb:BatchGetItem",
                "dynamodb:GetRecords",
                "dynamodb:GetShardIterator",
                "dynamodb:Query",
                "dynamodb:GetItem",
                "dynamodb:Scan",
                "dynamodb:ConditionCheckItem",
                "dynamodb:BatchWriteItem",
                "dynamodb:PutItem",
                "dynamodb:UpdateItem",
                "dynamodb:DeleteItem",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "dynamodbtable650E77A7",
                    "Arn",
                  ],
                },
                Object {
                  "Ref": "AWS::NoValue",
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "WebSocketHandlerServiceRoleDefaultPolicyF180304B",
        "Roles": Array [
          Object {
            "Ref": "WebSocketHandlerServiceRole79B07D0E",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "dynamodbtable650E77A7": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "AttributeDefinitions": Array [
          Object {
            "AttributeName": "connectionId",
            "AttributeType": "S",
          },
        ],
        "BillingMode": "PAY_PER_REQUEST",
        "KeySchema": Array [
          Object {
            "AttributeName": "connectionId",
            "KeyType": "HASH",
          },
        ],
        "TableName": "chat-app-table",
      },
      "Type": "AWS::DynamoDB::Table",
      "UpdateReplacePolicy": "Delete",
    },
    "s3staticwebsitebucket87408974": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "WebsiteConfiguration": Object {
          "IndexDocument": "index.html",
        },
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Delete",
    },
    "s3staticwebsitebucketPolicy607647E0": Object {
      "Properties": Object {
        "Bucket": Object {
          "Ref": "s3staticwebsitebucket87408974",
        },
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "s3:GetObject",
              "Effect": "Allow",
              "Principal": "*",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    Object {
                      "Fn::GetAtt": Array [
                        "s3staticwebsitebucket87408974",
                        "Arn",
                      ],
                    },
                    "/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
    },
    "s3staticwebsitedeploymentAwsCliLayerAA8A2B35": Object {
      "Properties": Object {
        "Content": Object {
          "S3Bucket": Object {
            "Ref": "AssetParameterse9882ab123687399f934da0d45effe675ecc8ce13b40cb946f3e1d6141fe8d68S3BucketAEADE8C7",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameterse9882ab123687399f934da0d45effe675ecc8ce13b40cb946f3e1d6141fe8d68S3VersionKeyE415415F",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameterse9882ab123687399f934da0d45effe675ecc8ce13b40cb946f3e1d6141fe8d68S3VersionKeyE415415F",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Description": "/opt/awscli/aws",
      },
      "Type": "AWS::Lambda::LayerVersion",
    },
    "s3staticwebsitedeploymentCustomResource9701CC61": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "DestinationBucketName": Object {
          "Ref": "s3staticwebsitebucket87408974",
        },
        "Prune": true,
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756C81C01536",
            "Arn",
          ],
        },
        "SourceBucketNames": Array [
          Object {
            "Ref": "AssetParameters450574362050a6ab3f053713b81bf81caf62329157f069c8c6eb58698a03b1c2S3BucketED4A2582",
          },
        ],
        "SourceObjectKeys": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters450574362050a6ab3f053713b81bf81caf62329157f069c8c6eb58698a03b1c2S3VersionKey1E02BCAA",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters450574362050a6ab3f053713b81bf81caf62329157f069c8c6eb58698a03b1c2S3VersionKey1E02BCAA",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        ],
      },
      "Type": "Custom::CDKBucketDeployment",
      "UpdateReplacePolicy": "Delete",
    },
    "websocketapi45E8C8EB": Object {
      "Properties": Object {
        "Name": "web-socket-api",
        "ProtocolType": "WEBSOCKET",
        "RouteSelectionExpression": "$request.body.action",
      },
      "Type": "AWS::ApiGatewayV2::Api",
    },
    "websocketapiconnectRoute7A1BE58E": Object {
      "Properties": Object {
        "ApiId": Object {
          "Ref": "websocketapi45E8C8EB",
        },
        "RouteKey": "$connect",
        "Target": Object {
          "Fn::Join": Array [
            "",
            Array [
              "integrations/",
              Object {
                "Ref": "websocketapiconnectRouteWebSocketIntegration6dfd9d8c896b35e88e44b9a3e07afd8129544D3C",
              },
            ],
          ],
        },
      },
      "Type": "AWS::ApiGatewayV2::Route",
    },
    "websocketapiconnectRouteWebSocketIntegration6dfd9d8c896b35e88e44b9a3e07afd8129544D3C": Object {
      "Properties": Object {
        "ApiId": Object {
          "Ref": "websocketapi45E8C8EB",
        },
        "IntegrationType": "AWS_PROXY",
        "IntegrationUri": Object {
          "Fn::Join": Array [
            "",
            Array [
              "arn:aws:apigateway:",
              Object {
                "Ref": "AWS::Region",
              },
              ":lambda:path/2015-03-31/functions/",
              Object {
                "Fn::GetAtt": Array [
                  "WebSocketHandler47C0AA1A",
                  "Arn",
                ],
              },
              "/invocations",
            ],
          ],
        },
      },
      "Type": "AWS::ApiGatewayV2::Integration",
    },
    "websocketapidefaultRoute3D6A725B": Object {
      "Properties": Object {
        "ApiId": Object {
          "Ref": "websocketapi45E8C8EB",
        },
        "RouteKey": "$default",
        "Target": Object {
          "Fn::Join": Array [
            "",
            Array [
              "integrations/",
              Object {
                "Ref": "websocketapiconnectRouteWebSocketIntegration6dfd9d8c896b35e88e44b9a3e07afd8129544D3C",
              },
            ],
          ],
        },
      },
      "Type": "AWS::ApiGatewayV2::Route",
    },
    "websocketapidisconnectRoute2127AFC1": Object {
      "Properties": Object {
        "ApiId": Object {
          "Ref": "websocketapi45E8C8EB",
        },
        "RouteKey": "$disconnect",
        "Target": Object {
          "Fn::Join": Array [
            "",
            Array [
              "integrations/",
              Object {
                "Ref": "websocketapiconnectRouteWebSocketIntegration6dfd9d8c896b35e88e44b9a3e07afd8129544D3C",
              },
            ],
          ],
        },
      },
      "Type": "AWS::ApiGatewayV2::Route",
    },
    "websocketapistageAFBBF77F": Object {
      "Properties": Object {
        "ApiId": Object {
          "Ref": "websocketapi45E8C8EB",
        },
        "AutoDeploy": true,
        "StageName": "prod",
      },
      "Type": "AWS::ApiGatewayV2::Stage",
    },
  },
}
`;
